crates/sui-graphql-rpc/src/types/cursor.rs

```diff
@@ -180,6 +184,31 @@ impl<C> Page<C> {
     }
 }
 
+impl<C> Page<C>
+where
+    C: Checkpointed,
+{
+    // If cursors are provided, defer to the `checkpoint_viewed_at` in the cursor if they are
+    // consistent. Otherwise, use the value from the parameter, or set to None. This is so that
+    // paginated queries are consistent with the previous query that created the cursor.
+    pub(crate) fn validate_cursor_consistency(&self) -> Result<Option<u64>, Error> {
+        match (self.after(), self.before()) {
+            (Some(after), Some(before)) => {
+                if after.checkpoint_viewed_at() == before.checkpoint_viewed_at() {
+                    Ok(after.checkpoint_viewed_at())
+                } else {
+                    Err(Error::Client(
+                        "The provided cursors are taken from different checkpoints and cannot be used together in the same query."
+                            .to_string(),
+                    ))
+                }
+            }
+            (Some(cursor), None) | (None, Some(cursor)) => Ok(cursor.checkpoint_viewed_at()),
```

Comment:
Will leave this comment at the final resting place of this code -- should this case be considered inconsistent?

Reply:
Handled at the top of the stack where both cursors if not-null will have non-null checkpoint_viewed_at values - will be considered inconsistent if the values are different